app/src/ui/lib/ref-name-text-box.tsx

```diff
@@ -0,0 +1,176 @@
+import * as React from 'react'
+
+import { sanitizedRefName } from '../../lib/sanitize-ref-name'
+import { TextBox } from './text-box'
+import { Octicon, OcticonSymbol } from '../octicons'
+import { Ref } from './ref'
+
+interface IRefNameProps {
+  /**
+   * The initial value for the ref name.
+   *
+   * Note that updates to this prop will be ignored.
```

Comment:
Gosh this irks me a bit but I'm not gonna hold up the PR for it. I can see how it'd be messy to try and have the parent component keep track of both the sanitized and the "raw" value but it sure doesn't feel idiomatic. I think I'd  feel a tad bit better if we at least reacted to changes of this prop but :shrug:

Reply:
The main problem about reacting to changes to this prop is that then we would have two sources of truth for the state: one that comes from the upwards props and another that comes from what the user types.

This could cause problems if the component gets re-rendered while/after the user has changed the input value: the props won't match the state anymore and whatever the user has typed will get lost (this whole thing is way better explained [in this article](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#common-bugs-when-using-derived-state)).

The other option would be to make this component controlled and move the state handling to its parents. As you mention this would add some additional boilerplate there and would leak the abstraction around the sanitization (this wouldn't be a big deal but I think that the current solution is slightly better).

Also, if we did that, note that we would just move this pattern to each of this component parents, making them uncontrolled (in fact, they are already uncontrolled, e.g `<CreateBranchDialog />` that accepts a `initialName` prop which gets ignored in subsequent re-renders ([source](https://github.com/desktop/desktop/blob/development/app/src/ui/create-branch/create-branch-dialog.tsx#L45)).

Anyways, if you feel strongly about it I can move the state that controls the raw value upwards, but imho this will only move the "problem" to a few places instead of having it in a single component (although is not a big change so I'm open to do it to unblock this)

Reply:
Yeah, I understand the tradeoffs and no, I don't feel strongly enough about it to hold up this PR. It's a pragmatic solution to a pre-existing condition.